### PR TITLE
[bugfix] Fix abnormal grad_norm under GRPO LoRA + DeepSpeed ZeRO-0 (fix #6815)

### LINKED ISSUE #6815: [Bug] Abnormal `grad_norm` in GRPO LoRA training with DeepSpeed ZeRO-0 (Normal with ZeRO-3)
**Describe the bug**
I encountered an issue where the `grad_norm` becomes abnormally large (e.g., ~1656) when performing GRPO training with LoRA using **DeepSpeed ZeRO-0**. However, when switching to **DeepSpeed ZeRO-3**, the `grad_norm` returns to a normal range (e.g., ~0.025).

This behavior suggests a potential issue with gradient synchronization or calculation specific to the ZeRO stage configuration in the GRPO LoRA pipeline.

**Your hardware and system info**
* **Hardware:** 4x A800
* **ms-swift Version:** 3.10.2
* **DeepSpeed Version:** Verified on both 0.17.6 and 0.18.x
* **Model:** Qwen3-8B
* **Training Type:** LoRA (Rank 16)
* **Task:** RLHF (GRPO)

**Reproduction Steps**
1.  **Model:** Qwen3-8B
2.  **Training Method:** GRPO + LoRA (Rank 16)
3.  **Loss Type:** BNPO
4.  **Configuration:**
    * **Case A (Bug):** Run the training command with `deepspeed` configured to ZeRO-0.
    * **Case B (Normal):** Run the same command with `--deepspeed zero3`.

**Reproduction Script:**
```bash
# 4x 80G GPUs
CUDA_VISIBLE_DEVICES=0,1,2,3 \
NPROC_PER_NODE=4 \
swift rlhf \
    --rlhf_type grpo \
    --model_type qwen3 \
    --model <path_to_model> \
    --dataset <path_to_dataset> \
    --external_plugins <plugins> \
    --reward_funcs <funcs> \
    --overlong_filter true \
    --max_completion_length 2048 \
    --learning_rate 1e-6 \
    --torch_dtype bfloat16 \
    --num_train_epochs 1 \
    --per_device_train_batch_size 4 \
    --per_device_eval_batch_size 4 \
    --save_total_limit 32 \
    --save_steps 50 \
    --logging_steps 2 \
    --output_dir output_retro_multi_grpo_3 \
    --gradient_accumulation_steps 32 \
    --warmup_steps 50 \
    --dataloader_num_workers 4 \
    --num_generations 8 \
    --use_vllm true \
    --vllm_gpu_memory_utilization 0.3 \
    --vllm_max_model_len 2560 \
    --sleep_level 0 \
    --offload_model false \
    --offload_optimizer false \
    --gc_collect_after_offload true \
    --tensor_parallel_size 1 \
    --temperature 0.8 \
    --top_p 0.85 \
    --log_completions true \
    --loss_type bnpo \
    --epsilon_high 0.28 \
    --dynamic_sample true \
    --max_resample_times 4 \
    --deepspeed zero0 \
    --train_type lora \
    --target_modules all-linear \
    --lora_rank 16
```
*(Note: Change `--deepspeed` to `zero3` for normal behavior)*

**Logs Comparison**

**Abnormal Log (DeepSpeed ZeRO-0):**
Note the `grad_norm`: **1656.27**
```json
{
  "loss": 0.01327927,
  "grad_norm": 1656.27481113,
  "learning_rate": 1e-07,
  "completions/mean_length": 830.90234375,
  "completions/min_length": 544.0,
  "completions/max_length": 1752.0,
  "completions/clipped_ratio": 0.0,
  "reward": 1.45605469,
  "reward_std": 0.4302305,
  "frac_reward_zero_std": 0.0,
  "rewards/RetroReward/mean": 1.45605469,
  "rewards/RetroReward/std": 0.51162845,
  "kl": 0.0,
  "clip_ratio/low_mean": 0.0,
  "clip_ratio/low_min": 0.0,
  "clip_ratio/high_mean": 0.0,
  "clip_ratio/high_max": 0.0,
  "clip_ratio/region_mean": 0.0,
  "epoch": 6.895e-05,
  "global_step/max_steps": "1/14503",
  "percentage": "0.01%",
  "elapsed_time": "2m 55s",
  "remaining_time": "29d 10h 53m 37s",
  "memory(GiB)": 52.96,
  "train_speed(iter/s)": 0.005699
}
```

**Normal Log (DeepSpeed ZeRO-3):**
Note the `grad_norm`: **0.025**
```json
{
  "loss": 0.01327927,
  "grad_norm": 0.02527279,
  "learning_rate": 1e-07,
  "completions/mean_length": 830.90234375,
  "completions/min_length": 544.0,
  "completions/max_length": 1752.0,
  "completions/clipped_ratio": 0.0,
  "reward": 1.45605469,
  "reward_std": 0.4302305,
  "frac_reward_zero_std": 0.0,
  "rewards/RetroReward/mean": 1.45605469,
  "rewards/RetroReward/std": 0.51162845,
  "kl": 0.0,
  "clip_ratio/low_mean": 0.0,
  "clip_ratio/low_min": 0.0,
  "clip_ratio/high_mean": 0.0,
  "clip_ratio/high_max": 0.0,
  "clip_ratio/region_mean": 0.0,
  "epoch": 6.895e-05,
  "global_step/max_steps": "1/14503",
  "percentage": "0.01%",
  "elapsed_time": "3m 20s",
  "remaining_time": "33d 16h 7m 29s",
  "memory(GiB)": 52.23,
  "train_speed(iter/s)": 0.004985
}
```

**Additional context**
* The issue seems specifically tied to the DeepSpeed stage configuration.
* The loss values are identical in the logs provided, but the gradient norm differs significantly, indicating that the forward pass might be consistent, but the backward pass or gradient aggregation is behaving differently under ZeRO-0.

### PR DESCRIPTION
**Pull request by alphadl**
https://github.com/modelscope/ms-swift/issues/6815

Under **DeepSpeed ZeRO-0** (or plain DDP), the logged `grad_norm` was abnormally large (e.g. ~1656) compared to **ZeRO-3** (e.g. ~0.025) with the same loss. This made monitoring and debugging difficult and was reported by multiple users.

**Root cause:** With ZeRO-0/DDP, each rank may report a different gradient norm (e.g. local view before or after reduce). ZeRO-3 reports a global norm. The trainer was logging the per-rank value without reduction, so the logged `grad_norm` was inconsistent across ZeRO stages.

**Changes:**
- `swift/trainers/mixin.py`:
  - Add `_get_reduced_grad_norm_for_logging(grad_norm)` to reduce `grad_norm` across processes when not ZeRO-3 and `world_size > 1` (all-reduce with `ReduceOp.AVG`), so that the logged value is consistent and comparable across ZeRO stages.
  - In `_maybe_log_save_evaluate`, use `_get_reduced_grad_norm_for_logging(args[0])` when building `logs['grad_norm']` (transformers >= 4.38).
- `tests/train/test_grad_norm_reduce.py`: Add unit tests for `_get_reduced_grad_norm_for_logging` (None/float/tensor, single-process, ZeRO-3 no reduce, ZeRO-0 with all_reduce).

**Behavior:**
- Single process or ZeRO-3: no change (existing behavior).
- Multi-process + ZeRO-0/1/2: `grad_norm` is all-reduced (average) before logging, so logs show a single consistent value.

### CODE REVIEW

**Comment by alphadl** [comment]
Please take a look when you have time. This fixes abnormal `grad_norm` logging under GRPO LoRA + DeepSpeed ZeRO-0 by all-reducing `grad_norm` across processes when not ZeRO-3 (fixes #6815). cc @Jintao-Huang @hjh0119
